### ISSUE TITLE
Java Keywords

### ISSUE THREAD

**Issue by zhongqi1112** [zhongqi1112/Blog#6]
# Introduce Important Java Keywords

- [static](http://zhongqi1112.me/2018/09/17/Java-Keywords/#static)
- [final](http://zhongqi1112.me/2018/09/17/Java-Keywords/#final)
- abstract
- interface
- [Scope](http://zhongqi1112.me/2018/09/17/Java-Keywords/#scope)

## static

Static is a non-access modifier in Java which is applicable for the following:

- blocks
- variables
- methods
- nested classes

When a member(block, variable, method, nested class) is declared static, it can be accessed before any objects of its class are created, and without reference to any object.

### Static blocks

If you need to do the computation in order to initialize your static variables, you can declare a static block that gets **executed exactly once**, when the first time you make an object of that class or the first time you access a static member of that class (even if you never make an object of that class). Also, static blocks are **executed before constructors**.

### Static variables

When a variable is declared as static, then a single copy of the variable is created and **shared among all objects at class level**. Static variables are, essentially, global variables. All instances of the class share the same static variable.

**Important points for static variables:**

- We can create static variables at the class-level only. Unlike C/C++, **static local variables are not allowed** in Java.
- Static block and static variables are executed in the order they are present in a program.

### Static methods

When a method is declared with a static keyword, it is known as a static method. The most common example of a static method is **main( ) method**. As discussed above, Any static member can be accessed before any objects of its class are created, and without reference to any object. Methods declared as static have several **restrictions**:

- They can only directly call other static methods.
- They can only directly access static data.
- They cannot refer to this or super in any way.

**When to use static variables and methods?**

Use the static variable for the property that is **common to all objects**. For example, in class Student, students share the same college name. Use static methods for changing static variables.

**Can a class be static in Java ?**
The answer is YES, we can have a static class in java. In java, we have static instance variables as well as static methods and also static blocks. Classes can also be made static in Java.

Java allows us to define a class within another class. Such a class is called a nested class. The class which enclosed the nested class is known as the Outer class. In java, we can’t make Top-level class static. **Only nested classes can be static**.

## Nested Classes

n java, it is possible to define a class within another class, such classes are known as *nestedclasses*. They enable you to logically group classes that are only used in one place, thus this increases the use of encapsulation, and create more readable and maintainable code.

- The scope of a nested class is bounded by the scope of its enclosing class. Thus in the above example, class *NestedClass* does not exist independently of class *OuterClass*.
- A nested class has access to the members, including private members, of the class in which it is nested. However, the reverse is not true i.e. the enclosing class does not have access to the members of the nested class.
- A nested class is also a member of its enclosing class.
- As a member of its enclosing class, a nested class can be declared *private*, *public*, *protected*, or *package private*(default).
- Nested classes are divided into two categories:
  1. **static nested class:** Nested classes that are declared *static* are called static nested classes.
  2. **inner class:** An inner class is a non-static nested class.

**Syntax:**

```Java
class OuterClass
{
...
    class NestedClass
    {
        ...
    }
}
```

### Static nested classes

As with class methods and variables, a static nested class is associated with its outer class. And like static class methods, a static nested class cannot refer directly to instance variables or methods defined in its enclosing class: it can use them only through an object reference.
They are accessed using the enclosing class name.

```Java
OuterClass.StaticNestedClass
```

For example, to create an object for the static nested class, use this syntax:

```Java
OuterClass.StaticNestedClass nestedObject =
     new OuterClass.StaticNestedClass();
```

### Inner classes

To instantiate an inner class, you must first instantiate the outer class. Then, create the inner object within the outer object with this syntax:

```Java
OuterClass.InnerClass innerObject = outerObject.new InnerClass();
```

**There are two special kinds of inner classes:**

1. Local inner classes
2. Anonymous inner classes

**What are the differences between static and non-static nested (inner) classes?** 
The following are major differences between the static nested class and non-static nested class. A non-static nested class is also called Inner Class.

**1)** Nested static class doesn’t need a reference to Outer class, but Non-static nested class or Inner class requires Outer class reference.

**2)** Inner class(or non-static nested class) can access both static and non-static members of the Outer class. A static class cannot access non-static members of the Outer class. It can access only static members of the Outer class.

**3)** An instance of Inner class cannot be created without an instance of the outer class and an Inner class can reference data and methods defined in Outer class in which it nests, so we don’t need to pass the reference of an object to the constructor of the Inner class. For this reason, Inner classes can make the program simple and concise.

## final

*final* keyword is used in different contexts. First of all, *final* is a non-access modifier applicable **only to a variable, a method, or a class**. The following are different contexts where the final is used.

### Final variables

When a variable is declared with *final* keyword, its value **can’t be modified**, essentially, a constant. This also means that you **must initialize a final variable**. If the final variable is a reference, this means that the variable cannot be re-bound to reference another object, but an internal state of the object pointed by that reference variable can be changed i.e. you can add or remove elements from a final array or final collection. It is good practice to represent final variables in all uppercase, using underscore to separate words.

**Initializing a final variable:**
We must initialize a final variable, otherwise, the compiler will throw compile-time error. A final variable can only be initialized once, either via an initializer or an assignment statement. There are three ways to initialize a final variable :

1. You can initialize a final variable when it is declared. This approach is the most common. A final variable is called **blank final variable**,if it is **not** initialized while declaration. Below are the two ways to initialize a blank final variable.
2. A blank final variable can be initialized inside the instance-initializer block or inside the constructor. If you have more than one constructor in your class then it must be initialized in all of them, otherwise, a compile-time error will be thrown.
3. A blank final static variable can be initialized inside a static block.

Let us see above different ways of initializing a final variable through an example.

```Java
//Java program to demonstrate differently 
// ways of initializing a final variable 
  
class Gfg  
{ 
    // a final variable 
    // direct initialize 
    final int THRESHOLD = 5; 
      
    // a blank final variable 
    final int CAPACITY; 
      
    // another blank final variable 
    final int  MINIMUM; 
      
    // a final static variable PI 
    // direct initialize 
    static final double PI = 3.141592653589793; 
      
    // a  blank final static  variable 
    static final double EULERCONSTANT; 
      
    // instance initializer block for  
    // initializing CAPACITY 
    { 
        CAPACITY = 25; 
    } 
      
    // static initializer block for  
    // initializing EULERCONSTANT 
    static{ 
        EULERCONSTANT = 2.3; 
    } 
      
    // constructor for initializing MINIMUM 
    // Note that if there are more than one 
    // constructor, you must initialize MINIMUM 
    // in them also 
    public GFG()  
    { 
        MINIMUM = -1; 
    } 
} 
```

**When to use a final variable:**

The only difference between a normal variable and a final variable is that we can re-assign value to a normal variable but we cannot change the value of a final variable once assigned. Hence final variables must be used only for the values that we want to remain constant throughout the execution of the program.

**Reference final variable:** 
When a final variable is a reference to an object, then this final variable is called reference final variable. For example, a final StringBuffer variable looks like

```
final StringBuffer sb;
```

As you know that a final variable cannot be re-assign. But in the case of a reference final variable, the internal state of the object pointed by that reference variable can be changed. Note that this is not re-assigning. This property of *final* is called *non-transitivity*.

The *non-transitivity* property also applies to arrays, because arrays are objects in the java. Arrays with the final keyword are also called final arrays.

**Note:**

1. As discussed above, a final variable cannot be reassigned.

2. When a final variable is created inside a method/constructor/block, it is called the local final variable, and it must initialize once where it is created. See the below program for the local final variable.

3. Note the difference between C++ *const* variables and Java *final* variables. const variables in C++ must be assigned a value when declared. For final variables in Java, it is not necessary as we see in the above examples. A final variable can be assigned value later, but only once.

4. *final* with foreach loop: final with for-each statement is a legal statement.

   **Explanation:** Since the i variable goes out of scope with each iteration of the loop, it is actually re-declaration each iteration, allowing the same token (i.e. i) to be used to represent multiple variables.

### Final classes

When a class is declared with *final* keyword, it is called a final class. A final class **cannot be extended(inherited)**. There are two uses of a final class :

1. One is definitely to **prevent inheritance**, as final classes cannot be extended. For example, all Wrapper Classes like Integer, Float, etc. are final classes. We can not extend them.
2. The other use of final with classes is to create an immutable class like the predefined String class. You can not make a class immutable without making it final.

### Final methods

When a method is declared with *final* keyword, it is called a final method. A final method **cannot be overridden**. The Object class does this—a number of its methods are final. We must declare methods with the final keyword for which we required to follow the same implementation throughout all the derived classes. 

## Scope

The scope of a variable is the part of the program where the variable is accessible.

### Member variables

These must be declared inside the class (outside any function). They can be directly accessed anywhere in the class. 

- We can declare class variables anywhere in class, but outside methods.
- Access specified of member variables doesn’t affect the scope of them within a class.
- Member variables can be accessed outside a class with following rules

```shell
Modifier      Package  Subclass  World

public          Yes      Yes     Yes

protected       Yes      Yes     No

Default (no
modifier)       Yes       No     No

private         No        No     No
```

### Local Variables (Method Level Scope)

Variables declared inside a method have the method-level scope and can’t be accessed outside the method.

**Note:** Local variables don’t exist after the method’s execution is over. Except the variable got passed in as a parameter to the method

### Loop Variables (Block Scope)

A variable declared inside pair of brackets “{” and “}” in a method has scope within the brackets only.

**Some Important Points about Variable scope in Java:**

- In general, a set of curly brackets { } defines a scope.
- In Java, we can usually access a variable as long as it was defined within the same set of brackets as the code we are writing or within any curly brackets inside of the curly brackets where the variable was defined.
- Any variable defined in a class outside of any method can be used by all member methods.
- When a method has the same local variable as a member, this keyword can be used to reference the current class variable.
- For a variable to be read after the termination of a loop, It must be declared before the body of the loop.

# License

[MIT License](./LICENSE) Copyright (c) 2018 By [Stephen Wang](http://zhongqi1112.me)

**References**:

- [Static keyword in java](https://www.geeksforgeeks.org/static-keyword-java/)
- [Static blocks in Java](https://www.geeksforgeeks.org/g-fact-79/)
- [Static class in Java](https://www.geeksforgeeks.org/static-class-in-java/)
- [Nested Classes in Java](https://www.geeksforgeeks.org/nested-classes-java/)
- [final keyword in java](https://www.geeksforgeeks.org/final-keyword-java/)
- [abstract keyword in java](https://www.geeksforgeeks.org/abstract-keyword-in-java/)
- [Scope of Variables In Java](https://www.geeksforgeeks.org/variable-scope-in-java/)
